src/null-safety/migration-guide.md

```diff
@@ -1,23 +1,472 @@
 ---
-title: Null safety migration guide
-description: Watch this space for information on how to make your code null safe
-toc: false
+title: Migrating to null safety
+description: How to move your existing Dart code to the world of null safety
 ---
 
-When the migration tool for null safety is available,
-this page will have information on how to
-change your production code to be null safe.
-Until then, you're welcome to try out null safety,
-but please don't depend on it.
+This page describes how and when to migrate your code to [null safety][].
+Here are the basic steps for migrating each package that you own:
 
-{{site.alert.important}}
-  Watch the [Dart announcements][] group and the [Dart blog][]
-  for announcements about null safety.
-{{site.alert.end}}
+1. [**Wait**](#step1-wait) for the packages
+   that you depend on to migrate.
+2. [**Migrate**](#step2-migrate) your package's code,
+   preferably using the interactive migration tool.
+3. [**Statically analyze**](#step3-analyze) your package's code.
+4. [**Test**](#step4-test) to make sure your changes work.
+5. If the package is already on pub.dev,
+   [**publish**](#step5-publish) the null-safe version
+   as a **prerelease** version.
 
-For more information, see the [null safety homepage][].
+{{ site.alert.info }}
+  **Migrating an app is technically the same as migrating a package.**
+  Before migrating an app,
+  consider waiting until null safety is in a stable release
+  and all your dependencies are ready.
+{{ site.alert.end }}
 
-[null safety homepage]: /null-safety
-[Dart announcements]: {{site.group}}/d/forum/announce
-[Dart blog]: https://medium.com/dartlang
+[null safety]: /null-safety
 
+![Screenshot of migration tool](/null-safety/migration-tool.png)
+_The interactive migration tool can simplify migration to null safety._
+
+
+## 1. Wait to migrate {#step1-wait}
+
+We strongly recommend migrating code in order, 
+with the leaves of the dependency graph being migrated first.
+For example, if package C depends on package B, which depends on package A,
+then A should be migrated to null safety first, then B, then C.
+
+![Illustration of C/B/A sentence](/null-safety/null-safety-migration-order.png){:width="454px"}<br>
+
+Although you _can_ migrate before your dependencies support null safety,
+you might have to change your code when your dependencies migrate.
+For example, if you predict that a function will take a nullable parameter but
+the package migrates it to be non-nullable,
+then passing a nullable argument becomes a compile error.
```

Comment:
**nit**: Is it perhaps worth pointing out that to run code migrated out-of-order you'll have to:
* (A) have an entrypoint that is opted-out of null-safety (possibly using a _language-version-marker_, e.g. add `// @dart=2.7` to `main.dart`), or,
 * (B) Run the code with `dart --no-sound-null-safety main.dart`.

For this reason package authors are strongly discouraged from publishing packages migrated out-of-order.
Since the downstream consumers of the published packages will be surprised if they have to pass `--no-sound-null-safety`.

This might be the wrong place to write this, maybe there is an FAQ where we can give a longer explanation of the downsides to migration out-of-order (perhaps with small examples of what can go wrong). AFAIK reasons are:
 * (i) `null` values may be able to sneak in places where they are not allowed, trigger run-time errors that wouldn't be possible in a full null-safety code base.
 * (ii) semantics of your dependencies will likely change as they migrate, potentially breaking your code.
 * (iii) users of your package will get a degraded experience because they have to pass `--no-sound-null-safety` (and they will suffer (i) and (ii), which is why this flag is required, as a warning to them).

Note: (iii) is unimportant if you're developing an application, but if you are migrating a package published on `pub.dev` this is a very good reason to postpone migration until dependencies are migrated.

Reply:
@kwalrath I think this was intended to be covered in https://github.com/dart-lang/site-www/pull/2764? If so, we could perhaps link to that from 'can migrate'?

Reply:
We should probably be more explicit about some of these downsides in #2764, but mit's right, that's where we intend to cover this.